### DISCUSSION TITLE
Why my DynamicMethod always return error

### DISCUSSION THREAD

**Discussion by MasterBran** [Excel-DNA/ExcelDna#453] [Q&A]
I want to create user-defined formula in Excel addin by ExcelDna ,and before I get the api response I don't know the name of the formula.All of the formula is just like that(It's worked good):
```
public static object GETROOT(object Param1, object Param2)
        {
            GetApiResult("001659ca6e3ed001d9a9be59d903a15c", new object[]
            {
                Param1,
                Param2
            });
            return "LOADING";
        }
```
The `GetApiResult` is an asynchronous function to get some data from server,and the fist parameter is a string.Based on that limits,I tried to create a `DynamicMethod` to reach my goal,this is my code:
```
public static void UDFCreate(Detail detail)
        {
            #region 
            string APIID = detail.apiIndDetail.id; // 
            string name = detail.apiIndDetail.formula;// 
            string description = detail.apiIndDetail.name;//
            List<string> paramList = new List<string>();// 
            var inputList = detail.apiIndInputs; // 
            foreach (var input in inputList)
            {
                paramList.Add(input.code);
            }
            Type[] types = new Type[paramList.Count];
            foreach (string param in paramList)
            {
                int index = paramList.IndexOf(param);
                types[index] = typeof(object);
            }
            #endregion
            DynamicMethod mult = new DynamicMethod(name, typeof(object), types, typeof(UDFGenerator));
            ILGenerator il = mult.GetILGenerator();
            //EmitTwo(il, paramList, APIID);
            il.Emit(OpCodes.Nop);
            il.Emit(OpCodes.Ldarg_0);
            il.Emit(OpCodes.Ldstr, APIID);
            il.Emit(OpCodes.Ldc_I4, paramList.Count);
            il.Emit(OpCodes.Newarr, typeof(object));
            foreach (var param in paramList)
            {
                int index = paramList.IndexOf(param);
                il.Emit(OpCodes.Dup);
                il.Emit(OpCodes.Ldc_I4, index);
                il.Emit(OpCodes.Ldarg_S, index + 1);
                il.Emit(OpCodes.Stelem_Ref);
            }
            MethodInfo mi = typeof(UDFGenerator).GetMethod(nameof(UDFGenerator.GetApiResult));
            il.Emit(OpCodes.Call, mi);
            il.Emit(OpCodes.Nop);
            il.Emit(OpCodes.Ldstr, "LOADING");
            il.Emit(OpCodes.Ret);
            List<Delegate> delegates = new List<Delegate>();            
            delegates.Add(mult.CreateDelegate(typeof(Func<object, object, object>)));
            
            ExcelFunctionAttribute att = new ExcelFunctionAttribute
            {
                Name = name,
                Description = description,
                IsMacroType = true,
            };
            List<object> funcAttribs = new List<object>
            {
                att
            };
            
            List<object> argAttribs = new List<object>();
            foreach (var input in inputList)
            {
                ExcelArgumentAttribute atta = new ExcelArgumentAttribute();
                atta.Name = input.name;
                argAttribs.Add(atta);
            }

            List<List<object>> argAttribsList = new List<List<object>>();
            argAttribsList.Add(argAttribs);
            try
            {
                ExcelIntegration.RegisterDelegates(delegates, funcAttribs, argAttribsList);
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }
        }
```
Finally,the formula is created, but when i use it in excel, there always return the error `A value used in the formula in the wrong data type`

**Comment by govert** [comment]
Can you add a definition and an example "Detail" object that I should try to run the code with?
I'm guessing it's something small that doesn't match up, but it would be easier to see if I can run it in the debugger.

**Reply by MasterBran** [reply]
> Can you add a definition and an example "Detail" object that I should try to run the code with? I'm guessing it's something small that doesn't match up, but it would be easier to see if I can run it in the debugger.
 The definition of "Detail" object
`public class Detail
    {
        public ApiIndDetail apiIndDetail { get; set; }
        public List<ApiIndInputsItem> apiIndInputs { get; set; }
        public List<ApiIndOutputsItem> apiIndOutputs { get; set; }
        public List<ApiIndMappingItem> apiIndMapping { get; set; }
    }
    public class ApiIndDetail : ObservableObject
    {
        public string id { get; set; }
        public string createBy { get; set; }
        public string createAt { get; set; }
        public string updateBy { get; set; }
        public string updateAt { get; set; }
        public string enableFlag { get; set; }
        public int sort { get; set; }
        public string catalogId { get; set; }
        public string isActv { get; set; }
        public string name { get; set; }
        public string catalogName { get; set; }
        public string catalogFullName { get; set; }
        public string formula { get; set; }
        public string typ { get; set; }
        public string typName { get; set; }
        public string remark { get; set; }
        public string apiId { get; set; }
        public string apiName { get; set; }
        public int vsNo { get; set; }
        public string indVerId { get; set; }
        public string apiInputIds { get; set; }
        public string apiOutputIds { get; set; }
        public string tagIds { get; set; }
        public string apiVsId { get; set; }
    }
    public class ApiIndInputsItem : ObservableObject
    {
        public string id { get; set; }
        public string code { get; set; }
        public string name { get; set; }
        public string inputTyp { get; set; }
        public string defVal { get; set; }
        public string defValOth { get; set; }
        public string isDisp { get; set; }
        public string notNull { get; set; }
        public string preProc { get; set; }
        public string srcTyp { get; set; }
        public string srcDefFirst { get; set; }
        public string source { get; set; }
        public string srcPara { get; set; }
        public string srcManuCont { get; set; }
        public string indVerId { get; set; }
    }
    public class ApiIndOutputsItem : ObservableObject
    {
        public string id { get; set; }
        public string code { get; set; }
        public string name { get; set; }
        public string apiDataTyp { get; set; }
        public string outputId { get; set; }
        public string dataTyp { get; set; }
        public string dataUnitTyp { get; set; }
        public string dataUnitTypName { get; set; }
        public string dataUnit { get; set; }
        public string dataUnitName { get; set; }
        public string apiDataUnit { get; set; }
        public string apiDataUnitName { get; set; }
        public string prcs { get; set; }
        public string prcsName { get; set; }
        public string isPerMil { get; set; }
        public string aligin { get; set; }
        public string aliginName { get; set; }
        public string style { get; set; }
        public string preProc { get; set; }
        public string indVerId { get; set; }
    }
    public class ApiIndMappingItem
    {
        public string id { get; set; }
        public string inputCode { get; set; }
        public string inputId { get; set; }
        public string indInputCode { get; set; }
        public string dateTyp { get; set; }
        public string indVerId { get; set; }
    }`
An example "Detail" object
"Detail": {
    "apiIndDetail": {
      "id": "001659ca6e3ed001d9a9be59d903a15c",
      "createBy": "1",
      "createAt": "2022-04-21 06:05:54",
      "updateBy": "1",
      "updateAt": "2022-04-25 03:15:27",
      "enableFlag": "1",
      "sort": 1,
      "catalogId": "3df46d1062462bc9addd6791b35b47f6",
      "isActv": "1",
      "name": "getroot-WILL",
      "catalogName": "WILLs",
      "catalogFullName": "WILLs",
      "formula": "GETROOT",
      "typ": "2",
      "typName": "table",
      "remark": "",
      "apiId": "d4e5b3cfec8d46b33de909384a123412",
      "apiName": "root",
      "vsNo": 11,
      "indVerId": "21632149665bd6f6e9756ae6af105ff3",
      "apiInputIds": "b5a64ff9b65e8a7327e6c7ce75268986,9b34e0fa89cd1e615bb325aadd558efe",
      "apiOutputIds": "d58e93bc3957e4b957d6a25df3f2a931,678bdb3da568b6d244e5363897096995,f0bb565da16d75634113f17d183dfa8f",
      "tagIds": null,
      "apiVsId": null,
      "authRoleNm": "CFO"
    },
    "apiIndInputs": [
      {
        "id": "b105bb5daba2bb87435bceeca10a7fd2",
        "code": "NM",
        "name": "firstName",
        "inputTyp": "4",
        "defVal": "",
        "defValOth": null,
        "isDisp": "1",
        "notNull": "0",
        "preProc": null,
        "srcTyp": null,
        "srcDefFirst": null,
        "source": "",
        "srcPara": null,
        "srcManuCont": null,
        "indVerId": "21632149665bd6f6e9756ae6af105ff3"
      },
      {
        "id": "2c23ba933daba19ba1dbcfb9adf46e39",
        "code": "ENM",
        "name": "secondName",
        "inputTyp": "1",
        "defVal": "",
        "defValOth": null,
        "isDisp": "1",
        "notNull": "0",
        "preProc": null,
        "srcTyp": null,
        "srcDefFirst": null,
        "source": null,
        "srcPara": null,
        "srcManuCont": null,
        "indVerId": "21632149665bd6f6e9756ae6af105ff3"
      }
    ],
    "apiIndOutputs": [
      {
        "id": "caef5c755e8093b75922a4c798991307",
        "code": "ID",
        "name": "ID",
        "apiDataTyp": null,
        "outputId": "d58e93bc3957e4b957d6a25df3f2a931",
        "dataTyp": "2",
        "dataUnitTyp": "TEXT",
        "dataUnitTypName": "test",
        "dataUnit": null,
        "dataUnitName": null,
        "apiDataUnit": null,
        "apiDataUnitName": null,
        "prcs": null,
        "prcsName": null,
        "isPerMil": "1",
        "aligin": "1",
        "aliginName": "left",
        "style": null,
        "preProc": null,
        "indVerId": "21632149665bd6f6e9756ae6af105ff3"
      },
      {
        "id": "ff70986081a8ffb7032515bc53d6266a",
        "code": "ENG_FN",
        "name": "firstName",
        "apiDataTyp": null,
        "outputId": "678bdb3da568b6d244e5363897096995",
        "dataTyp": "1",
        "dataUnitTyp": "TEXT",
        "dataUnitTypName": "test",
        "dataUnit": null,
        "dataUnitName": null,
        "apiDataUnit": null,
        "apiDataUnitName": null,
        "prcs": null,
        "prcsName": null,
        "isPerMil": "1",
        "aligin": "1",
        "aliginName": "left",
        "style": null,
        "preProc": null,
        "indVerId": "21632149665bd6f6e9756ae6af105ff3"
      },
      {
        "id": "9fb6c920b26760ada720311faba67bbb",
        "code": "ENG_ABBR",
        "name": "shortName",
        "apiDataTyp": null,
        "outputId": "f0bb565da16d75634113f17d183dfa8f",
        "dataTyp": "1",
        "dataUnitTyp": "TEXT",
        "dataUnitTypName": "test",
        "dataUnit": null,
        "dataUnitName": null,
        "apiDataUnit": null,
        "apiDataUnitName": null,
        "prcs": null,
        "prcsName": null,
        "isPerMil": "1",
        "aligin": "1",
        "aliginName": "left",
        "style": null,
        "preProc": null,
        "indVerId": "21632149665bd6f6e9756ae6af105ff3"
      }
    ],
    "apiIndMapping": [
      {
        "id": "0d1ab8aa62feaeb365d5a62de80d983b",
        "inputCode": "NAME",
        "inputId": "b5a64ff9b65e8a7327e6c7ce75268986",
        "indInputCode": "NM",
        "dateTyp": null,
        "indVerId": "21632149665bd6f6e9756ae6af105ff3"
      },
      {
        "id": "3bce3f5113d40eff820ecbc927727f7d",
        "inputCode": "ENG_FN",
        "inputId": "d35351664fbde82f38b1f79e1effc5c2",
        "indInputCode": "ENM",
        "dateTyp": null,
        "indVerId": "21632149665bd6f6e9756ae6af105ff3"
      }
    ]
  }
